lib/ecto/query/planner.ex

```diff
@@ -442,16 +442,23 @@ defmodule Ecto.Query.Planner do
   end
 
   defp rewrite_join(%{on: on, ix: join_ix} = join, qual, ix, last_ix, source_ix, inc_ix) do
-    on = update_in on.expr, fn expr ->
-      Macro.prewalk expr, fn
-        {:&, meta, [join_ix]} ->
-          {:&, meta, [rewrite_ix(join_ix, ix, last_ix, source_ix, inc_ix)]}
-        other ->
-          other
-      end
-    end
+    rewritten_on =
+      on
+      |> Map.update!(:expr, fn expr ->
+        Macro.prewalk expr, fn
+          {:&, meta, [join_ix]} ->
+            {:&, meta, [rewrite_ix(join_ix, ix, last_ix, source_ix, inc_ix)]}
+          expr = %Ecto.Query.Tagged{type: {type_ix, type}} when is_integer(type_ix) ->
+            %{expr | type: {rewrite_ix(type_ix, ix, last_ix, source_ix, inc_ix), type}}
+          other ->
+            other
+        end
+      end)
+      |> Map.update!(:params, fn params ->
+        Enum.map(params, &rewrite_param_ix(&1, ix, last_ix, source_ix, inc_ix))
+      end)
```

Comment:
I'd prefer to have this written without `Map.update!`
```elixir
%{expr: expr, params: params} = on
expr = 
  ...
params =
  ...

%{join | on: %{on | expr: expr, params: params}, qual: qual ...}
```

I think it is slightly clearer what's happening.